lib/ansible/plugins/inventory/__init__.py

```diff
@@ -230,12 +230,12 @@ def _read_config_data(self, path):
         if not config:
             # no data
             raise AnsibleParserError("%s is empty" % (to_native(path)))
+        elif not isinstance(config, Mapping):
+            # configs are dictionaries
+            raise AnsibleParserError('inventory source has an invalid structure, it should be a dictionary, got: %s' % type(config))
         elif config.get('plugin') not in valid_names:
```

Comment:
This code expects `config` as a dictionary. Moved the `isinstance` check above this line to make sure that `config` is always a dictionary.

Reply:
that depends on the plugin, it should not be required by the common code